### PR TITLE
feat: replace oauth lambda with cfn

### PR DESCRIPTION
**Pull request by akshbhu**
#### Description of changes
- removes the OAUTH custom resource which sets  OAUTH properties on userPool client
- added CFN logic on userpoolClient  resource to  add OAUTH properties directly

#### Description of how you validated changes
- manually tested
- added migration test
- green pipeline : https://app.circleci.com/pipelines/github/aws-amplify/amplify-cli/18788/workflows/cef793a9-9dca-4bc8-9659-402ac18d963c

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#tests)
- [ ] Relevant documentation is changed or added (and PR referenced)
- [ ] New AWS SDK calls or CloudFormation actions have been added to relevant test and service IAM policies
- [ ] [Pull request labels](https://github.com/aws-amplify/amplify-cli/blob/dev/CONTRIBUTING.md#labels) are added

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by Amplifiyer** [comment]
Is this a `fix` or a `feat`? Can you also add more details in the description of the PR?

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts [hunk 1]
```diff
@@ -646,6 +649,36 @@ export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCog
...
+  updateUserPoolClientWithOAuthSettings = (props: CognitoStackOptions): void => {
+    const oAuthMetaData = JSONUtilities.parse<OAuthMetaData>(props.oAuthMetadata);
+    let hostedUIProviderMeta;
+    let supportedIdentityProviders = [];
+    if (!_.isEmpty(props.hostedUIProviderMeta)) {
+      hostedUIProviderMeta = JSONUtilities.parse<Array<any>>(props.hostedUIProviderMeta);
+      supportedIdentityProviders = hostedUIProviderMeta.map((provider: { ProviderName: any }) => provider.ProviderName);
+    }
+    supportedIdentityProviders.push('COGNITO');
+    if (this.userPoolClient) {
```
Comment (Amplifiyer) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:664]: Is this a valid state? i.e. can `userPoolClient` ever be null/undefined here?

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts [hunk 2]
```diff
@@ -646,6 +649,36 @@ export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCog
...
+      this.userPoolClient.allowedOAuthFlowsUserPoolClient = true;
+      this.userPoolClient.allowedOAuthScopes = oAuthMetaData?.AllowedOAuthScopes;
+      this.userPoolClient.allowedOAuthFlows = oAuthMetaData?.AllowedOAuthFlows;
+      this.userPoolClient.callbackUrLs = oAuthMetaData?.CallbackURLs;
+      this.userPoolClient.logoutUrLs = oAuthMetaData?.LogoutURLs;
+      this.userPoolClient.supportedIdentityProviders = supportedIdentityProviders;
+    }
+    if (this.userPoolClientWeb) {
```
Comment (Amplifiyer) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:672]: Same as above, is it possible for this to be false and if yes, is that a valid state?

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts [hunk 3]
```diff
@@ -646,6 +649,36 @@ export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCog
     });
   }
 
+  /**
+   * creates OAuth customResource for Cognito
```
Comment (Amplifiyer) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:653]: We are not creating a `customResource`, right? But updating the Cognito's L1 construct with OAuth?

**Review by Amplifiyer — COMMENTED** [review]
Probably a dumb question, but I didn't see anything getting "replaced" in this PR. I thought we are replacing a custom resource lambda with this?

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts (hunk 1 above)
Comment (akshbhu) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:664]: no , I did that to remove non null `!` checks.

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts (hunk 3 above)
Comment (akshbhu) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:653]: yeah you are correct , the previous commit on this branch : https://github.com/aws-amplify/amplify-cli/commit/a391baf5cb0241052638e780a737691e08f4d9bf#diff-79c5ca11b4f77a8d94fb923276bd9beea861fa0d543c188627496b3b79939ad7 removes the OAuthcustom resource.

I can update the comment though

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts (hunk 2 above)
Comment (akshbhu) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:672]: Similar, it cant  be undefined here if Cx selected to create userPool

**Review by Amplifiyer — APPROVED** [review]

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts [hunk 4]
```diff
@@ -646,6 +649,36 @@ export class AmplifyAuthCognitoStack extends cdk.Stack implements AmplifyAuthCog
...
+    if (!_.isEmpty(props.hostedUIProviderMeta)) {
+      hostedUIProviderMeta = JSONUtilities.parse<Array<$TSAny>>(props.hostedUIProviderMeta);
+      supportedIdentityProviders = hostedUIProviderMeta.map((provider: { ProviderName: string }) => provider.ProviderName);
+    }
+    supportedIdentityProviders.push('COGNITO');
+    if (this.userPoolClient) {
+      this.userPoolClient.allowedOAuthFlowsUserPoolClient = true;
+      this.userPoolClient.allowedOAuthScopes = oAuthMetaData?.AllowedOAuthScopes;
+      this.userPoolClient.allowedOAuthFlows = oAuthMetaData?.AllowedOAuthFlows;
+      this.userPoolClient.callbackUrLs = oAuthMetaData?.CallbackURLs;
```
Comment (jhockett) [inline packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:668]: nit: casing `callbackUrLs` => `callbackURLs`

File: packages/amplify-migration-tests/src/migration-helpers-v11/auth-helpers/utilities.ts [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+    appleAppPrivateKey:
+      '----BEGIN PRIVATE KEY----MIGTAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBHkwdwIBAQQgIltgNsTgTfSzUadYiCS0VYtDDMFln/J8i1yJsSIw5g+gCgYIKoZIzj0DAQehRANCAASI8E0L/DhR/mIfTT07v3VwQu6q8I76lgn7kFhT0HvWoLuHKGQFcFkXXCgztgBrprzd419mUChAnKE6y89bWcNw----END PRIVATE KEY----',
+  };
+};
+
+/**
+ *  generates a uuid
+ */
+
+export const getShortId = (): string => {
```
Comment (jhockett) [inline packages/amplify-migration-tests/src/migration-helpers-v11/auth-helpers/utilities.ts:40]: `generateRandomShortId` which is exported from `@aws-amplify/amplify-e2e-core` accomplishes the same thing.

File: packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts (hunk 4 above)
Comment (akshbhu) [reply packages/amplify-category-auth/src/provider-utils/awscloudformation/auth-stack-builder/auth-cognito-stack-builder.ts:668]: its coming from here: https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.aws_cognito.CfnUserPoolClient.html#callbackurls

**Review by danielleadams — APPROVED** [review]
LGTM

**Pull request merged**